### DISCUSSION TITLE
How to build on M1 chip Mac?

### DISCUSSION THREAD

**Discussion by huangyingw** [RsyncProject/rsync#404] [Q&A]
I used following command to build, not familiar with c/c++, trying to add include path to the compiler.

```
brew install \
    automake \
    lz4 \
    openssl \
    xxhash \
    zstd

export LDFLAGS="-L$HOMEBREW_PREFIX/opt/openssl@3/lib"
export CPPFLAGS="-I$HOMEBREW_PREFIX/opt/openssl@3/include"
export CPPFLAGS="-I$HOMEBREW_PREFIX/Cellar/xxhash/0.8.1/include"
export PKG_CONFIG_PATH="$HOMEBREW_PREFIX/opt/openssl@3/lib/pkgconfig"
export CPLUS_INCLUDE_PATH=/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include 
export DYLD_LIBRARY_PATH=/usr/local/include

./configure && \
    make && \
    sudo make install
```

But still failed with following errors:

```
Configure found the following issues:

- Failed to find openssl/md4.h and openssl/md5.h for openssl crypto lib support.
- Failed to find XXH64_createState function in xxhash lib.
- Failed to find zstd.h for zstd compression support.
- Failed to find lz4.h for lz4 compression support.
```

**Comment by WayneD** [comment]
Right after that text it tells you what configure options to use if you want to choose to skip those features. If you instead want to include those features (which is to be preferred), install those missing libraries:

https://download.samba.org/pub/rsync/INSTALL

The opening text talks about the libs, but for a crib sheet, scroll down to the macOS brew instructions.

**Reply by huangyingw** [reply]
I want to include those features. 
And I did follow the macos instruction, to use brew to install those deps, like openssl, ...etc.
But I still ended up with those "Failed to find *" configuration errors.

**Comment by WayneD** [comment]
Give the macOS [CI rules](https://github.com/WayneD/rsync/blob/master/.github/workflows/build.yml) a look and see if you see anything. That runs fine on an x86_64. I think it runs as root.

Your openssl error looks like it's finding LibreSSL instead of openssl, so the search order isn't working.

The only arm macOS I have access to doesn't have brew on it, so I just download the various tar files (for lz4, openssl, xxHash, and zstd) along with installing python3's commonmark via -mpip. With the tars extracted they build quite easily. I then put the .h files into an "include" dir and the .a files in a "lib" dir (both in the parent dir of my rsync dir):

```
include:
lz4.h                   lz4frame_static.h       openssl/                zdict.h                 zstd_errors.h
lz4frame.h              lz4hc.h                 xxhash.h                zstd.h

lib:
libcrypto.a     liblz4.a        libxxhash.a     libzstd.a
```

And then run configure like this:

```console
CFLAGS='-I../include' LDFLAGS='-L../lib' ./configure
```
I managed to add macOS 12.6 on arm64 to the available binaries, which may help you: https://download.samba.org/pub/rsync/binaries/macos-12.6-arm64/

**Reply by huangyingw** [reply]
great thanks, based on your input, the following command work in my m1 chip mac:
```
CPPFLAGS="-I$HOMEBREW_PREFIX/opt/openssl/include/ \
    -I$HOMEBREW_PREFIX/opt/xxhash/include/ \
    -I$HOMEBREW_PREFIX/opt/zstd/include/ \
    -I$HOMEBREW_PREFIX/opt/lz4/include/" \
    LDFLAGS="-L$HOMEBREW_PREFIX/opt/openssl/lib/ \
    -L$HOMEBREW_PREFIX/opt/xxhash/lib/ \
    -L$HOMEBREW_PREFIX/opt/zstd/lib/ \
    -L$HOMEBREW_PREFIX/opt/lz4/lib/"\
    ./configure \
    --prefix=$HOMEBREW_PREFIX && \
    make reconfigure && \
    make && \
    sudo make install
```
